### PR TITLE
Vsr individual

### LINKED ISSUE #11: VSR Individual View for Staff
Implement the VSR individual view page for staff. This should show one VSR in detail. Focus on the frontend first, because we need the VSR DB model to be completed in order to build the backend route for this task.

### Frontend
- Implement the VSR individual view, following the [Figma Design](https://www.figma.com/file/jARPdBWD6HutwLFzepjZW1/PAP-Web-Application?type=design&node-id=200-2520&mode=design&t=tO9klt7Btz8fMuUY-0)
- Put this page on the `/staff/vsr/[id]` URL (see the [NextJS dynamic routes documentation](https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes))
- Use the [Material UI Accordion component](https://mui.com/material-ui/react-accordion/) for the accordion sections
- You can use hard-coded dummy data for now, to test the view and ensure it matches the design
- Once the backend for this task is complete, the frontend should call the backend route to get the data for the individual VSR.

### Backend
- Once our VSR MongoDB database model is done, add an API route to GET `/api/vsr/[id]` to get data on a specific VSR. The `id` in the URL path should be the ID of the VSR (MongoDB assigns each DB entry a unique ID). This route should verify that the user is authenticated (using a Firebase token in the header) and has the `staff` or `admin` role. If these checks pass, retrieve the corresponding VSR and return its data in the response

### PR DESCRIPTION
**Pull request by Syu125**
## Tracking Info

Resolves #11 

## Changes

- Completed frontend of individual vsr and connected database to pull form values for display.

## Testing

- Checked formatting and interacted with page to make sure everything was functional.
- Created a data entry in MongoDB and modified values to check that it was displayed correctly on the page.

## Confirmation of Change

1. Start MongoDB server
2. Open up two terminals, one for backend and one for frontend, and make sure all the dependencies are installed
    - Backend: navigate to backend directory > `npm run start`
    - Frontend: navigate to frontend directory > `npm run dev`
3. [Open page](http://localhost:3000/staff/vsr/65bc31561826f0d6ee2c4b21)
    - *If values are not showing up, upload `frontend\__tests__\sampleData\pap.vsrs.json` to MongoDB*
5. You can modify the values in the database to test

### CODE REVIEW

File: frontend/src/components/Page.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/Page.tsx]: Can we make this text and the "Export" button text white to match the Figma? Right now, they're black

File: frontend/src/components/SingleDetail.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/SingleDetail.tsx]: Nice job with the email link!

File: frontend/src/components/MilitaryBackground.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/MilitaryBackground.tsx]: For this field and other array fields, it shows nothing at all if the value is `[]` (an empty array) because your code is showing "N/A" only if the value is set to `undefined`. Can we expand it to show "N/A" if the value is either "falsy" (undefined/null/false) or empty? So something like

<ListDetail title="Branch" values={(vsr.branch && vsr.branch.length > 0)  ? vsr.branch : ["N/A"]} />

Same thing for conflicts, discharge status, etc.

File: frontend/src/components/SingleDetail.module.css
Comment (benjaminJohnson2204) [inline frontend/src/components/SingleDetail.module.css]: Can you add `white-space: nowrap;` here so that the "Case ID" label doesn't wrap to the next line?

File: frontend/src/components/CaseDetails.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/CaseDetails.tsx]: For `dateReceived` and `lastUpdated`, can you format them like the Figma? I think the default format they're displayed in right now is long and harder to visually parse at a glance

File: frontend/src/components/Page.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/Page.tsx]: Can you make this link to `/staff/vsr`, which is where the VSR table view will be, instead of `complete`?

File: frontend/src/components/DropdownDetail.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/DropdownDetail.tsx]: If this code isn't being used, can you remove it?

File: frontend/src/components/Page.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/Page.tsx]: For the page parameters/props, I think you can just have this page not take any props. So get rid of `PageProps` and just have `export const Page = () => {`

File: frontend/src/components/PersonalInformation.module.css
Comment (benjaminJohnson2204) [inline frontend/src/components/PersonalInformation.module.css]: If this CSS class isn't being used, can you remove it?

Comment (benjaminJohnson2204) [reply frontend/src/components/PersonalInformation.module.css]: Same for the `accordianTitle` class in several other CSS files

File: frontend/src/components/RequestedFurnishings.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/RequestedFurnishings.tsx]: I noticed that each of your components is calling `getVSR` in order to get the VSR for the current page. This results in multiple unnecessary network requests for the same data. Instead, can we just have the top-level `Page` component call `getVSR` and then pass the VSR to the various other components (`ContactInfo`, `PersonalInformation`, etc.)?

File: frontend/src/components/RequestedFurnishings.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/RequestedFurnishings.tsx]: Also, in the one place we call `getVSR`, we should have some error handling. If the backend returns an error, we should display a corresponding error message to the user (e.g. network error, VSR not found). The way it is now, we still display the same VSR page but with none of the fields filled in, so it's hard to tell there's an error

**Review by benjaminJohnson2204 — CHANGES_REQUESTED** [review]
Thanks for getting this PR out, and great work! The page looks really nice and you got it working end-to-end - amazing job!

I left a few comments. Most of them are minor changes regarding code style, unused code, and CSS. 

I think since you merged Steven and Yoto's branch into yours to have the VSR DB model, I'll probably wait until Steven and Yoto open a PR before I merge your PR, since merging yours now would be merging some of Steven and Yoto's unfinished work. So no rush on addressing my comments.

Thanks again!

**Comment by Syu125** [comment]
> Thanks for getting this PR out, and great work! The page looks really nice and you got it working end-to-end - amazing job!
> 
> I left a few comments. Most of them are minor changes regarding code style, unused code, and CSS.
> 
> I think since you merged Steven and Yoto's branch into yours to have the VSR DB model, I'll probably wait until Steven and Yoto open a PR before I merge your PR, since merging yours now would be merging some of Steven and Yoto's unfinished work. So no rush on addressing my comments.
> 
> Thanks again!

Thank you! Yeah, sounds good. I'll try to get these things done before next meeting.

File: frontend/src/components/Page.tsx
Comment (Syu125) [reply frontend/src/components/Page.tsx]: I think the button texts are white already - it is a rendering issue then?
This is what I see:
image

File: frontend/src/components/SingleDetail.tsx
Comment (Syu125) [reply frontend/src/components/SingleDetail.tsx]: Thanks!

File: frontend/src/components/MilitaryBackground.tsx
Comment (Syu125) [reply frontend/src/components/MilitaryBackground.tsx]: Ok, I updated all the list ones, including those in the MilitaryBackground and PersonalInformation sections.

File: frontend/src/components/SingleDetail.module.css
Comment (Syu125) [reply frontend/src/components/SingleDetail.module.css]: Do you mean that the "Case ID" label in the Case Details section? What do you mean by it's wrapping?
This is what I see:
image

File: frontend/src/components/CaseDetails.tsx
Comment (Syu125) [reply frontend/src/components/CaseDetails.tsx]: This is what I see - do they look like this on your end?
image

File: frontend/src/components/Page.tsx
Comment (Syu125) [reply frontend/src/components/Page.tsx]: Yep!

File: frontend/src/components/Page.tsx
Comment (Syu125) [reply frontend/src/components/Page.tsx]: Ok, I removed it.

File: frontend/src/components/DropdownDetail.tsx
Comment (Syu125) [reply frontend/src/components/DropdownDetail.tsx]: Yeah

File: frontend/src/components/PersonalInformation.module.css
Comment (Syu125) [reply frontend/src/components/PersonalInformation.module.css]: Yep, just cleaned it up.

File: frontend/src/components/RequestedFurnishings.tsx
Comment (Syu125) [reply frontend/src/components/RequestedFurnishings.tsx]: Sounds good - I moved getVSR into Page.

File: frontend/src/components/RequestedFurnishings.tsx
Comment (Syu125) [reply frontend/src/components/RequestedFurnishings.tsx]: I added an error handling, so an error will be displayed where the name would be. Would we want a modal in the future?
image

File: frontend/src/components/Page.tsx
Comment (benjaminJohnson2204) [reply frontend/src/components/Page.tsx]: This is what it looks like for me. Maybe your browser has white as the default text or something. Setting the font color to white explicitly should fix it
Screenshot from 2024-02-10 11-04-25

File: frontend/src/components/SingleDetail.module.css
Comment (benjaminJohnson2204) [reply frontend/src/components/SingleDetail.module.css]: It looks like this for me, not sure why it looks different for you vs me. Adding the nowrap property should fix it though
Screenshot from 2024-02-10 11-05-28

File: frontend/src/components/CaseDetails.tsx
Comment (benjaminJohnson2204) [reply frontend/src/components/CaseDetails.tsx]: Oh I see, I think the dates are formatted differently on my end because they were saved in a different format in my local database, and you're just displaying the string from the database. We should change these fields to be dates instead of strings later, but for now this is fine

File: frontend/src/components/RequestedFurnishings.tsx
Comment (benjaminJohnson2204) [reply frontend/src/components/RequestedFurnishings.tsx]: Ok, this is good for now. We might want a different way of displaying errors in the future, we can talk to the designers. But can you change it so if there's an error, we only display the error message and don't display the other VSR components?

File: frontend/src/components/MilitaryBackground.tsx
Comment (benjaminJohnson2204) [inline frontend/src/components/MilitaryBackground.tsx]: Can you remove this commented-out code too?

File: frontend/src/components/Page.tsx
Comment (Syu125) [reply frontend/src/components/Page.tsx]: Ahh ok. Fixed it.

File: frontend/src/components/SingleDetail.module.css
Comment (Syu125) [reply frontend/src/components/SingleDetail.module.css]: Ok, added it.

File: frontend/src/components/MilitaryBackground.tsx
Comment (Syu125) [reply frontend/src/components/MilitaryBackground.tsx]: Yep

**Review by benjaminJohnson2204 — APPROVED** [review]
LGTM! I made some refactoring changes to reduce the amount of re-used code across the different components, but great job!

**Review by 2s2e — APPROVED** [review]
-Got this warning while using mongosh, not sure if relevant: DeprecationWarning: Collection.insert() is deprecated. Use insertOne, insertMany, or bulkWrite.
-Ages display seems to be clumped together 
image
This is for ages [2,9,5]

Aside from the minor points above, everything looks good! The website closely follows the figma and there weren't any visual bugs. Changes to the VSR in MongoDB were updated on the website.

**Comment by benjaminJohnson2204** [comment]
Thanks for the prompt review @2s2e ! I wouldn't worry about the the `mongosh` warning. Nice catch with the ages display, I just pushed a fix.

**Review by HarshGurnani — COMMENTED** [review]
The date is showing up as invalid, is this something that needs to be changed? It may be because at `frontend/src/api/VSRs.ts`, the `VSRJson` type uses `string` for the two dates rather than a `Date` type, so a string representation that is unrecognized as a date is stored in the Mongo. 

image

**Review by HarshGurnani — APPROVED** [review]
lgtm!

**Pull request merged**
